File: src/main/scala/tlschannel/TlsSocketChannel.scala
package tlschannel

import java.nio.ByteBuffer
import java.nio.channels.SocketChannel
import java.nio.channels.ByteChannel
import javax.net.ssl.SSLEngine
import javax.net.ssl.SSLSession

trait TlsSocketChannel extends ByteChannel {

  def wrapped: ByteChannel
  def read(dstBuffer: ByteBuffer): Int
  def write(srcBuffer: ByteBuffer): Int

  def renegotiate()
  def doPassiveHandshake()
  def doHandshake(): Unit
  def close(): Unit
  
  def isOpen() = wrapped.isOpen

  def getSession(): SSLSession
  
}

File: src/main/java/tlschannel/TlsSocketChannel.java
package tlschannel;

import java.io.IOException;
import java.nio.channels.ByteChannel;
import java.nio.channels.GatheringByteChannel;
import java.nio.channels.ScatteringByteChannel;

import javax.net.ssl.SSLSession;

public interface TlsSocketChannel extends ByteChannel, GatheringByteChannel, ScatteringByteChannel {

	ByteChannel getWrapped();

	void renegotiate() throws IOException;

	void negotiate() throws IOException;

	void close();

	boolean isOpen();

	SSLSession getSession();

	boolean getRunTasks();
	
}
